File: observer/python/observable_product.py
# Created by Sezer BOZKIR<admin@sezerbozkir.com> at 22.12.2021
from product_update_message import ProductUpdateMessage
from abstract_observer_user import AbstractObserverUser


class ObservableProduct:
    """
    Takip edilecek olan yapımızdır.
    İçerisinde Observer (değişimden etkilenecek) nesneleri tutar.
    UML diyagramındaki Observable yapısına denk gelmektedir.
    """

    def __init__(self):
        self._users = list()  # SYB: list() yerine direkt "[]" şeklinde tanımlama da yapılabilir.
        self.message = ProductUpdateMessage()

    def add_observer(self, observer_user: AbstractObserverUser):
        self._users.append(observer_user)

    def remove_observer(self, observer_user: AbstractObserverUser):
        self._users.remove(observer_user)

    def notify_observer(self):
        for user in self._users:  # type:AbstractObserverUser
            user.send_notification(self.message)

File: observer/typescript/observable-product.ts
import {IObserverUser} from "./observer-user.interface";
import {ProductUpdateMessage} from "./product-update-message";

// Takip edilecek olan yapımızdır.
// İçerisinde Observer (değişimden etkilenecek) nesneleri tutar.
// UML diyagramındaki Observable yapısına denk gelmektedir.
export abstract class ObservableProduct {
  private users = new Set<IObserverUser>();
  protected message = new ProductUpdateMessage();

  public addObserver(observerUser: IObserverUser): void {
    this.users.add(observerUser);
  }

  public removeObserver(observerUser: IObserverUser): void {
    this.users.delete(observerUser);
  }

  public notifyObserver(): void {
    for (let user of this.users) {
      user.sendNotification(this.message);
    }
  }
}
